publications/radstack/radstack.tex

```diff
@@ -0,0 +1,1148 @@
+production clusters running at Metamarkets in early 2015. We also include
+results from synthetic workloads on TPC-H data.
+
+\subsection{Query Performance in Production}
+Druid query performance can vary signficantly depending on the query being
+issued. For example, sorting the values of a high cardinality dimension based
+on a given metric is much more expensive than a simple count over a time range.
+To showcase the average query latencies in a production Druid cluster, we
+selected 8 of our most queried data sources, described in
+Table~\ref{tab:datasources}.
+
+Approximately 30\% of queries are standard aggregates involving different types
+of metrics and filters, 60\% of queries are ordered group bys over one or more
+dimensions with aggregates, and 10\% of queries are search queries and metadata
+retrieval queries. The number of columns scanned in aggregate queries roughly
+follows an exponential distribution. Queries involving a single column are very
+frequent, and queries involving all columns are very rare.
+
+\begin{table}
+  \centering
+  \scriptsize\begin{tabular}{| l | l | l |}
+    \hline
+    \textbf{Data Source} & \textbf{Dimensions} & \textbf{Metrics} \\ \hline
+    \texttt{a} & 25 & 21 \\ \hline
+    \texttt{b} & 30 & 26 \\ \hline
+    \texttt{c} & 71 & 35 \\ \hline
+    \texttt{d} & 60 & 19 \\ \hline
+    \texttt{e} & 29 & 8 \\ \hline
+    \texttt{f} & 30 & 16 \\ \hline
+    \texttt{g} & 26 & 18 \\ \hline
+    \texttt{h} & 78 & 14 \\ \hline
+  \end{tabular}
+  \normalsize
+  \caption{Characteristics of production data sources.}
+  \label{tab:datasources}
+\end{table}
+
+\begin{itemize}[leftmargin=*,beginpenalty=5000,topsep=0pt]
+\item There were
+approximately 50 data sources in this particular cluster and several hundred users issuing
+queries.
+
+\item There was approximately 10.5TB of RAM available in this cluster and
+approximately 10TB of segments loaded. Collectively,
+there are about 50 billion Druid rows in this cluster. Results for
+every data source is not shown.
+
+\item This cluster uses Intel\textsuperscript{\textregistered} Xeon\textsuperscript{\textregistered} E5-2670 processors and consists of 1302 processing
+threads and 672 total cores (hyperthreaded).
+
+\item A memory-mapped storage engine was used (the machine was configured to
+    memory map the data instead of loading it into the Java heap.)
+\end{itemize}
+
+Query latencies are shown in Figure~\ref{fig:query_latency} and the queries per
+minute are shown in Figure~\ref{fig:queries_per_min}. Across all the various
+data sources, average query latency is approximately 550 milliseconds, with
+90\% of queries returning in less than 1 second, 95\% in under 2 seconds, and
+99\% of queries returning in less than 10 seconds. Occasionally we observe
+spikes in latency, as observed on February 19, where network issues on
+the Memcached instances were compounded by very high query load on one of our
```

Comment:
Memcached has not been mentioned yet, so "the Memcached instances" just comes out of nowhere.